atc/db/base_resource_type.go

```diff
@@ -1,11 +1,87 @@
 package db
 
 import (
-	"database/sql"
+	"sync"
+	"time"
 
 	sq "github.com/Masterminds/squirrel"
 )
 
+// The table base_resource_types is small (less than 100 tuples) and steady (it
+// may only be updated after Concourse changes). But it's queried extremely
+// frequently, thus it makes sense to cache this table in memory, and reload it
+// periodically, which should help reduce db queries.
```

Comment:
I keep running load tests recently to comparing this PR version with the master branch version. Some optimizations, like this one, are from the load test. I'll post some charts soon.

In my tests, the db is not super powerful. The DB is deployed on K8S with 8GHz cpu and 8GB memory. As each check build will run `step.resourceConfigFactory.FindOrCreateResourceConfig(step.plan.Type, source, resourceTypes)` and `delegate.FindOrCreateScope(resourceConfig)`, and I noticed many times they took seconds. Then I dumped slow queries from the db, and figured this is one of the slow query. Maybe because this query is executed too often.